cli/tfsource/types.go

```diff
@@ -45,21 +49,57 @@ func (src *TerraformSource) String() string {
 // so the same file path (/foo/bar) is always considered the same version. To detect changes the file path will be hashed
 // and returned as version. In case of hash error the default encoded source version will be returned.
 // See also the encodeSourceName and ProcessTerraformSource methods.
-func (terraformSource TerraformSource) EncodeSourceVersion() string {
+func (terraformSource TerraformSource) EncodeSourceVersion() (string, error) {
 	if IsLocalSource(terraformSource.CanonicalSourceURL) {
-		hash, err := dirhash.HashDir(terraformSource.CanonicalSourceURL.Path, "prefix", dirhash.DefaultHash)
+		sourceHash := sha256.New()
+		sourceDir := filepath.Clean(terraformSource.CanonicalSourceURL.Path)
+
+		err := filepath.Walk(sourceDir, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				// If we've encountered an error while walking the tree, give up
+				return err
+			}
+
+			if info.IsDir() {
+				// We don't use any info from directories to calculate our hash
+				return nil
+			}
+
+			fileModified := info.ModTime().UnixMicro()
+			hashContents := fmt.Sprintf("%s:%d", path, fileModified)
+
+			sourceHash.Write([]byte(hashContents))
+
+			return nil
+		})
+
 		if err == nil {
-			return hash
+			hash := fmt.Sprintf("%x", sourceHash.Sum(nil))
+
+			return hash, nil
+		} else {
+			terraformSource.Logger.WithError(err).Debugf("Could not encode version for local source")
+			return "", err
 		}
-		// In case of error return default source version strategy
 	}
-	return util.EncodeBase64Sha1(terraformSource.CanonicalSourceURL.Query().Encode())
+
+	return util.EncodeBase64Sha1(terraformSource.CanonicalSourceURL.Query().Encode()), nil
 }
 
 // Write a file into the DownloadDir that contains the version number of this source code. The version number is
 // calculated using the EncodeSourceVersion method.
 func (terraformSource TerraformSource) WriteVersionFile() error {
-	version := terraformSource.EncodeSourceVersion()
+	version, err := terraformSource.EncodeSourceVersion()
+	if err != nil {
+		// If we failed to calculate a SHA of the downloaded source, write a SHA of
+		// some random data into the version file.
+		//
+		// This ensures we attempt to redownload the source next time.
+		randomBytes := make([]byte, 32)
+		rand.Read(randomBytes)
```

Comment:
Maybe make sense to extract as separate function and re-use it in `alreadyHaveLatestCode()`  and `WriteVersionFile()`